// node-red-contrib-home-assistant-websocket/dist/common/controllers/SendSplitMixin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = sendSplitMixin;
const globals_1 = require("../../globals");
function sendSplitMixin(Base) {
    return class SendSplitController extends Base {
        sendSplit(message, data, send) {
            if (!send) {
                send = this.node.send;
            }
            delete message._msgid;
            message.parts = {
                id: globals_1.RED.util.generateId(),
                count: data.length,
                index: 0,
                // TODO: check if this works
                // type: 'array',
                // len: 1,
            };
            let pos = 0;
            for (let i = 0; i < data.length; i++) {
                message.payload = data.slice(pos, pos + 1)[0];
                if (message.parts) {
                    message.parts.index = i;
                }
                pos += 1;
                send(globals_1.RED.util.cloneMessage(message));
            }
        }
    };
}
